File: app/src/main/java/io/github/fvasco/pinpoi/importer/ImporterFacade.java
package io.github.fvasco.pinpoi.importer;

import android.app.ProgressDialog;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.net.URLConnection;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

import io.github.fvasco.pinpoi.BuildConfig;
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao;
import io.github.fvasco.pinpoi.dao.PlacemarkDao;
import io.github.fvasco.pinpoi.model.Placemark;
import io.github.fvasco.pinpoi.model.PlacemarkCollection;
import io.github.fvasco.pinpoi.util.Consumer;
import io.github.fvasco.pinpoi.util.ProgressDialogInputStream;
import io.github.fvasco.pinpoi.util.Util;

/**
 * Importer facade for:
 * <ul>
 * <li>KML</li>
 * <li>KMZ</li>
 * <li>GPX</li>
 * <li>RSS</li>
 * <li>OV2 Tomtom</li>
 * <li>ASC, CSV files</li>
 * </ul>
 * <p/>
 * Import {@linkplain Placemark} and update {@linkplain PlacemarkCollectionDao}
 *
 * @author Francesco Vasco
 */
public class ImporterFacade implements Consumer<Placemark> {

    /**
     * Signpost for end of elaboration
     */
    private final Placemark STOP_PLACEMARK = new Placemark();
    private final PlacemarkDao placemarkDao;
    private final PlacemarkCollectionDao placemarkCollectionDao;
    private final BlockingQueue<Placemark> placemarkQueue = new ArrayBlockingQueue<>(256);
    private ProgressDialog progressDialog;
    private String progressDialogMessageFormat;

    public ImporterFacade() {
        this(Util.getApplicationContext());
    }

    public ImporterFacade(Context context) {
        this.placemarkDao = new PlacemarkDao(context);
        this.placemarkCollectionDao = new PlacemarkCollectionDao(context);
    }

    @Nullable
    static AbstractImporter createImporter(@NonNull final String resource) {
        String path;
        try {
            if (Util.isUri(resource)) {
                final List<String> segments = Uri.parse(resource).getPathSegments();
                path = null;
                for (int i = segments.size() - 1; i >= 0 && Util.isEmpty(path); --i) {
                    path = segments.get(i);
                }
            } else {
                path = resource;
            }
        } catch (Exception e) {
            path = resource;
        }
        final AbstractImporter res;
        if (path == null || path.length() < 3) {
            res = null;
        } else {
            final String end = path.substring(path.length() - 3);
            switch (end.toLowerCase()) {
                case "kml":
                    res = new KmlImporter();
                    break;
                case "kmz":
                case "zip":
                    res = new ZipImporter();
                    break;
                case "xml":
                case "rss":
                    res = new GeoRssImporter();
                    break;
                case "gpx":
                    res = new GpxImporter();
                    break;
                case "ov2":
                    res = new Ov2Importer();
                    break;
                case "asc":
                case "csv":
                case "txt":
                    res = new TextImporter();
                    break;
                default:
                    res = null;
            }
        }
        Log.d(ImporterFacade.class.getSimpleName(),
                "Importer for " + resource + " is " + (res == null ? null : res.getClass().getSimpleName()));
        return res;
    }

    /**
     * Set optional progress dialog to show and update with progress
     *
     * @see #setProgressDialogMessageFormat(String)
     */
    public void setProgressDialog(ProgressDialog progressDialog) {
        this.progressDialog = progressDialog;
    }

    public void setProgressDialogMessageFormat(String progressDialogMessageFormat) {
        this.progressDialogMessageFormat = progressDialogMessageFormat;
    }

    /**
     * Import a generic resource into data base, this action refresh collection.
     * If imported count is 0 no modification is done.
     * <b>Side effect</b> update and save placermak collection
     *
     * @return imported {@linkplain io.github.fvasco.pinpoi.model.Placemark}
     */
    public int importPlacemarks(@NonNull final PlacemarkCollection placemarkCollection) throws IOException {
        final String resource = placemarkCollection.getSource();
        Objects.requireNonNull(resource, "Null source");
        placemarkCollectionDao.open();
        try {
            final AbstractImporter importer = createImporter(resource);
            if (importer == null) {
                throw new IOException("Cannot import " + resource);
            }
            importer.setCollectionId(placemarkCollection.getId());
            importer.setConsumer(this);

            // insert new placemark
            final Future<Void> importFuture = Util.EXECUTOR.submit(new Callable<Void>() {
                @Override
                public Void call() throws Exception {
                    InputStream inputStream;
                    final int max;
                    if (resource.startsWith("/")) {
                        final File file = new File(resource);
                        max = (int) file.length();
                        inputStream = new BufferedInputStream(new FileInputStream(file));
                    } else {
                        final URLConnection urlConnection = new URL(resource).openConnection();
                        inputStream = urlConnection.getInputStream();
                        max = urlConnection.getContentLength();
                    }
                    if (progressDialog != null) {
                        progressDialog.setIndeterminate(max <= 0);
                        progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
                        if (max > 0) {
                            progressDialog.setMax(max);
                            inputStream = new ProgressDialogInputStream(inputStream, progressDialog);
                        }
                        Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                            @Override
                            public void run() {
                                progressDialog.show();
                            }
                        });
                    }

                    try {
                        importer.importPlacemarks(inputStream);
                    } finally {
                        placemarkQueue.put(STOP_PLACEMARK);
                        inputStream.close();
                    }
                    return null;
                }
            });
            int placemarkCount = 0;
            placemarkDao.open();
            SQLiteDatabase placemarkDaoDatabase = placemarkDao.getDatabase();
            placemarkDaoDatabase.beginTransaction();
            try {
                // remove old placemark
                placemarkDao.deleteByCollectionId(placemarkCollection.getId());
                Placemark placemark;
                while ((placemark = placemarkQueue.take()) != STOP_PLACEMARK) {
                    try {
                        placemarkDao.insert(placemark);
                        ++placemarkCount;
                        if (progressDialog != null && progressDialogMessageFormat != null) {
                            final String message = String.format(progressDialogMessageFormat, placemarkCount);
                            Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                                @Override
                                public void run() {
                                    progressDialog.setMessage(message);
                                }
                            });
                        }
                    } catch (IllegalArgumentException e) {
                        // discard (duplicate?) placemark
                        if (BuildConfig.DEBUG) {
                            Log.i(ImporterFacade.class.getSimpleName(), "Placemark discarded " + placemark, e);
                        }
                    }
                }
                if (progressDialog != null) {
                    Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                        @Override
                        public void run() {
                            progressDialog.setIndeterminate(true);
                        }
                    });
                }
                // wait import and check exception
                importFuture.get();
                if (placemarkCount > 0) {
                    // update placemark collection
                    placemarkCollection.setLastUpdate(System.currentTimeMillis());
                    placemarkCollection.setPoiCount(placemarkCount);
                    placemarkCollectionDao.update(placemarkCollection);
                    // confirm transaction
                    placemarkDaoDatabase.setTransactionSuccessful();
                }
                return placemarkCount;
            } catch (InterruptedException | RuntimeException e) {
                throw new IOException("Error importing placemark", e);
            } catch (ExecutionException e) {
                throw new IOException("Error importing placemark", e.getCause());
            } finally {
                importFuture.cancel(true);
                placemarkDaoDatabase.endTransaction();
                placemarkDao.close();
            }
        } finally {
            placemarkCollectionDao.close();
            if (progressDialog != null) {
                Util.MAIN_LOOPER_HANDLER.post(new Runnable() {
                    @Override
                    public void run() {
                        progressDialog.dismiss();
                    }
                });
            }
        }
    }

    @Override
    public void accept(Placemark p) {
        try {
            placemarkQueue.put(p);
        } catch (InterruptedException e) {
            Log.w(ImporterFacade.class.getSimpleName(), "Placemark discarded " + p, e);
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/io/github/fvasco/pinpoi/importer/ImporterFacade.kt
package io.github.fvasco.pinpoi.importer

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.net.Uri
import android.util.Log
import io.github.fvasco.pinpoi.BuildConfig
import io.github.fvasco.pinpoi.dao.PlacemarkCollectionDao
import io.github.fvasco.pinpoi.dao.PlacemarkDao
import io.github.fvasco.pinpoi.model.Placemark
import io.github.fvasco.pinpoi.model.PlacemarkCollection
import io.github.fvasco.pinpoi.util.*
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.util.*
import java.util.concurrent.ArrayBlockingQueue

/**
 * Import [Placemark] and update [PlacemarkCollectionDao]
 *
 * Importer facade for:
 *  * KML
 *  * KMZ
 *  * GPX
 *  * RSS
 *  * OV2 Tomtom
 *  * ASC, CSV files
 * @author Francesco Vasco
 */
class ImporterFacade(context: Context) {

    private val placemarkDao: PlacemarkDao
    private val placemarkCollectionDao: PlacemarkCollectionDao
    private var progressDialog: ProgressDialog? = null
    private var progressDialogMessageFormat: String? = null
    var fileFormatFilter: FileFormatFilter = FileFormatFilter.NONE

    init {
        this.placemarkDao = PlacemarkDao(context)
        this.placemarkCollectionDao = PlacemarkCollectionDao(context)
    }

    /**
     * Set optional progress dialog to show and update with progress

     * @see .setProgressDialogMessageFormat
     */
    fun setProgressDialog(progressDialog: ProgressDialog) {
        this.progressDialog = progressDialog
    }

    fun setProgressDialogMessageFormat(progressDialogMessageFormat: String) {
        this.progressDialogMessageFormat = progressDialogMessageFormat
    }

    /**
     * Import a generic resource into data base, this action refresh collection.
     * If imported count is 0 no modification is done.
     * **Side effect** update and save placermak collection
     * @return imported [io.github.fvasco.pinpoi.model.Placemark]
     */
    @Throws(IOException::class)
    fun importPlacemarks(placemarkCollection: PlacemarkCollection, context: Context): Int {
        val resource = placemarkCollection.source
        val (mimeType, inputStream) = openInputStream(resource)
        val importer =
            if (ZipImporter.isZipInputStream(inputStream)) ZipImporter()
            else createImporter(resource, mimeType, fileFormatFilter)
                ?: throw IOException("Cannot import $resource")
        return importPlacemarks(placemarkCollection, importer, inputStream, context)
    }

    fun importPlacemarks(
        placemarkCollection: PlacemarkCollection,
        importer: AbstractImporter,
        inputStream: InputStream,
        context: Context
    ): Int {
        placemarkCollectionDao.open()
        val placemarkQueue = ArrayBlockingQueue<PlacemarkEvent>(64)
        try {
            runOnUiThread { progressDialog?.show() }

            // copy data locally to avoid network issue
            val tempFile = File(context.cacheDir, "importPlacemarks.tmp");
            tempFile.delete()
            tempFile.outputStream().buffered(8 * 1024).use(inputStream::copyTo)

            // insert new placemarks
            val importFuture = doAsync {
                try {
                    tempFile.inputStream().buffered().use { inputStream ->
                        try {
                            importer.collectionId = placemarkCollection.id
                            importer.consumer =
                                { placemarkQueue.put(PlacemarkEvent.NewPlacemark(it)) }
                            importer.importPlacemarks(inputStream)
                        } catch (e: Exception) {
                            placemarkQueue.put(PlacemarkEvent.ParseError(e))
                        }
                    }
                } finally {
                    placemarkQueue.put(PlacemarkEvent.End)
                }
            }
            var placemarkCount = 0
            placemarkDao.open()
            val placemarkDaoDatabase = placemarkDao.database!!
            placemarkDaoDatabase.beginTransaction()
            try {
                // remove old placemarks
                placemarkDao.deleteByCollectionId(placemarkCollection.id)
                var placemarkEvent = placemarkQueue.take()
                while (placemarkEvent is PlacemarkEvent.NewPlacemark) {
                    val placemark = placemarkEvent.placemark
                    if (placemarkDao.insert(placemark, SQLiteDatabase.CONFLICT_IGNORE)) {
                        ++placemarkCount
                    } else {
                        // discard (duplicate?) placemark
                        if (BuildConfig.DEBUG) {
                            Log.d(
                                ImporterFacade::class.java.simpleName,
                                "Placemark discarded $placemark"
                            )
                        }
                    }
                    placemarkEvent = placemarkQueue.take()
                }
                if (placemarkEvent is PlacemarkEvent.ParseError) throw placemarkEvent.throwable
                // wait import and check exception
                importFuture.get()
                if (placemarkCount > 0) {
                    // update placemark collection
                    placemarkCollection.lastUpdate = System.currentTimeMillis()
                    placemarkCollection.poiCount = placemarkCount
                    placemarkCollectionDao.update(placemarkCollection)
                    // confirm transaction
                    placemarkDaoDatabase.setTransactionSuccessful()
                }
                return placemarkCount
            } catch (e: Exception) {
                throw IOException("Error importing placemarks", e)
            } finally {
                importFuture.cancel(true)
                placemarkDaoDatabase.endTransaction()
                placemarkDao.close()
                tempFile.delete()
            }
        } finally {
            try {
                placemarkCollectionDao.close()
            } finally {
                runOnUiThread { progressDialog?.tryDismiss() }
            }
        }
    }

    companion object {

        private val supportedMimeType: SortedSet<String> =
            FileFormatFilter.values().flatMap { it.validMimeTypes }
                .plus("application/vnd.google-earth.kmz")
                .plus("application/zip").plus("application/x-zip")
                // mime type sorted ("application/" before "text/")
                .toSortedSet()

        internal fun createImporter(
            resource: String,
            mimeType: String?,
            fileFormatFilter: FileFormatFilter
        ): AbstractImporter? {
            var res: AbstractImporter? = null

            if (mimeType != null) res = createImporterFromMimeType(mimeType, fileFormatFilter)

            if (res == null) {
                val path = try {
                    if (resource.isUri()) Uri.parse(resource).pathSegments.lastOrNull() ?: resource
                    else resource
                } catch (e: Exception) {
                    resource
                }

                when (path.substringAfterLast('.').lowercase()) {
                    in FileFormatFilter.KML.validExtensions -> if (fileFormatFilter == FileFormatFilter.NONE || fileFormatFilter == FileFormatFilter.KML) res =
                        KmlImporter()

                    "kmz", "zip" -> res = ZipImporter()
                    "xml", "rss" -> if (fileFormatFilter == FileFormatFilter.NONE) res =
                        GeoRssImporter()

                    in FileFormatFilter.GPX.validExtensions -> if (fileFormatFilter == FileFormatFilter.NONE || fileFormatFilter == FileFormatFilter.GPX) res =
                        GpxImporter()

                    in FileFormatFilter.OV2.validExtensions -> if (fileFormatFilter == FileFormatFilter.NONE || fileFormatFilter == FileFormatFilter.OV2) res =
                        Ov2Importer()

                    in FileFormatFilter.CSV_LAT_LON.validExtensions -> if (fileFormatFilter == FileFormatFilter.NONE || fileFormatFilter == FileFormatFilter.CSV_LAT_LON) res =
                        TextImporter()

                    in FileFormatFilter.CSV_LON_LAT.validExtensions -> if (fileFormatFilter == FileFormatFilter.NONE || fileFormatFilter == FileFormatFilter.CSV_LON_LAT) res =
                        TextImporter()
                }
            }

            if (res == null) res = fileFormatFilter.toAbstractImporter()

            res?.fileFormatFilter = fileFormatFilter
            Log.d(
                ImporterFacade::class.java.simpleName,
                "Importer for " + resource + " (" + mimeType + ") is " + res?.javaClass?.simpleName
            )
            return res
        }

        fun createImporterFromMimeType(
            mimeType: String,
            fileFormatFilter: FileFormatFilter
        ): AbstractImporter? {
            when (mimeType.substringBefore(';')) {
                "application/json", "application/geo+json" -> return GeoJsonImporter()
            }
            if (mimeType.startsWith("text/")) return FileFormatFilter.CSV_LAT_LON.toAbstractImporter()

            val primarySubtype =
                mimeType.substringAfterLast('/').substringAfterLast('.').substringBefore('+')
            when (primarySubtype) {
                "kmz", "zip", "zip-compressed", "x-zip-compressed" -> return ZipImporter()
                "xml", "rss" -> return GeoRssImporter()
            }

            return (FileFormatFilter.values().find { primarySubtype in it.validExtensions }
                ?: fileFormatFilter)
                .toAbstractImporter()
        }

        private fun FileFormatFilter.toAbstractImporter(): AbstractImporter? =
            when (this) {
                FileFormatFilter.NONE -> null
                FileFormatFilter.CSV_LAT_LON, FileFormatFilter.CSV_LON_LAT -> TextImporter()
                FileFormatFilter.GEOJSON -> GeoJsonImporter()
                FileFormatFilter.GPX -> GpxImporter()
                FileFormatFilter.KML -> KmlImporter()
                FileFormatFilter.OV2 -> Ov2Importer()
                FileFormatFilter.RSS -> GeoJsonImporter()
            }

        private fun openInputStream(resource: String): Source {
            val connection = makeURL(resource).openConnection()
            connection.addRequestProperty(
                "Accept", supportedMimeType.joinToString(postfix = ", */*;q=0.1")
            )
            connection.connect()
            val mimeType: String? = connection.getHeaderField("Content-Type")
            val inputStream: InputStream = connection.getInputStream().buffered()
            return Source(mimeType, inputStream)
        }

        private data class Source(val mimeType: String?, val inputStream: InputStream)

        private sealed class PlacemarkEvent {
            class NewPlacemark(val placemark: Placemark) : PlacemarkEvent()
            object End : PlacemarkEvent()
            class ParseError(val throwable: Throwable) : PlacemarkEvent()
        }
    }
}
